analytics/src/main/java/com/segment/analytics/internal/Utils.java

```diff
@@ -302,40 +298,58 @@ public static String assertNotNullOrEmpty(String text, @Nullable String name) {
         return Collections.unmodifiableList(new ArrayList<>(list));
     }
 
-    /**
-     * Creates a unique device id. Suppresses `HardwareIds` lint warnings as we don't use this ID
-     * for identifying specific users. This is also what is required by the Segment spec.
-     */
-    @SuppressLint("HardwareIds")
-    public static String getDeviceId(Context context) {
-        String androidId = getString(context.getContentResolver(), ANDROID_ID);
-        if (!isNullOrEmpty(androidId)
-                && !"9774d56d682e549c".equals(androidId)
-                && !"unknown".equals(androidId)
-                && !"000000000000000".equals(androidId)) {
-            return androidId;
-        }
-
-        // Serial number, guaranteed to be on all non phones in 2.3+.
-        if (!isNullOrEmpty(Build.SERIAL)) {
-            return Build.SERIAL;
-        }
-
-        // Telephony ID, guaranteed to be on all phones, requires READ_PHONE_STATE permission
-        if (hasPermission(context, READ_PHONE_STATE) && hasFeature(context, FEATURE_TELEPHONY)) {
-            TelephonyManager telephonyManager = getSystemService(context, TELEPHONY_SERVICE);
-            @SuppressLint("MissingPermission")
-            String telephonyId = telephonyManager.getDeviceId();
-            if (!isNullOrEmpty(telephonyId)) {
-                return telephonyId;
-            }
+    /** Creates a unique device id. */
+    public static String getDeviceId() {
+        // unique id generated from DRM API
+        String uniqueID = getUniqueID();
+        if (!isNullOrEmpty(uniqueID)) {
+            return uniqueID;
         }
 
         // If this still fails, generate random identifier that does not persist across
         // installations
         return UUID.randomUUID().toString();
     }
 
+    /**
+     * Workaround for not able to get device id on Android 10 or above using DRM API {@see
+     * https://stackoverflow.com/questions/58103580/android-10-imei-no-longer-available-on-api-29-looking-for-alternatives}
+     * {@see https://developer.android.com/training/articles/user-data-ids}
+     */
+    private static String getUniqueID() {
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.JELLY_BEAN_MR2) return null;
```

Comment:
any reason to return null if we are lower than API 18?

Reply:
yes, the drm api is only available in 18+. so it's either we bump the min sdk to 18 or we do a check. and the `getDeviceID` method does a null check and returns random uuid as fallback.